src/runtime_src/core/tools/xbmgmt2/OO_UpdateXclbin.cpp

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2022 Advanced Micro Devices, Inc. All rights reserved.
+
+// ------ I N C L U D E   F I L E S -------------------------------------------
+// Local - Include Files
+#include "OO_UpdateXclbin.h"
+
+// XRT - Include Files
+#include "core/common/query_requests.h"
+#include "tools/common/XBUtilities.h"
+#include "tools/common/XBUtilitiesCore.h"
+namespace XBU = XBUtilities;
+
+// 3rd Party Library - Include Files
+#include <boost/format.hpp>
+#include <boost/program_options.hpp>
+namespace po = boost::program_options;
+
+// System - Include Files
+#include <fstream>
+#include <iostream>
+#include <vector>
+// =============================================================================
+
+// ----- H E L P E R M E T H O D S ------------------------------------------
+
+
+// ----- C L A S S   M E T H O D S -------------------------------------------
+
+OO_UpdateXclbin::OO_UpdateXclbin(const std::string &_longName, const std::string &_shortName, bool _isHidden )
+    : OptionOptions(_longName,
+                    _shortName,
+                    "Load an xclbin onto the FPGA",
+                    boost::program_options::value<decltype(xclbin)>(&xclbin)->required(),
+                    "The xclbin to be loaded.  Valid values:\n"
+                    "  Name (and path) of the xclbin.",
+                    _isHidden),
+      m_device(""),
+      xclbin("")
+{
+  m_optionsDescription.add_options()
+    ("device,d", po::value<decltype(m_device)>(&m_device), "The Bus:Device.Function (e.g., 0000:d8:00.0) device of interest")
+    ("help", po::bool_switch(&m_help), "Help to use this sub-command")
+  ;
+}
+
+void
+OO_UpdateXclbin::execute(const SubCmdOptions& _options) const
+{
+  XBUtilities::verbose("SubCommand option: Update xclbin");
+
+  XBUtilities::verbose("Option(s):");
+  for (const auto & aString : _options)
+    XBUtilities::verbose(std::string(" ") + aString);
+
+  // Honor help option first
+  if (std::find(_options.begin(), _options.end(), "--help") != _options.end()) {
+    printHelp();
+    return;
+  }
+
+  // Parse sub-command ...
+  po::variables_map vm;
+  process_arguments(vm, _options);
+
+  std::shared_ptr<xrt_core::device> device;
+  try {
+    device = XBU::get_device(boost::algorithm::to_lower_copy(m_device), false /*inUserDomain*/);
+  } catch (const std::runtime_error& e) {
+    XBU::throw_cancel(e.what());
+  }
+
+  XBU::sudo_or_throw("Root privileges are required to download xclbin");
+
+  std::ifstream stream(xclbin, std::ios::binary);
+  if (!stream)
+    throw xrt_core::error(boost::str(boost::format("Could not open %s for reading") % xclbin));
+
+  stream.seekg(0,stream.end);
+  ssize_t size = stream.tellg();
+  stream.seekg(0,stream.beg);
```

Comment:
Is there a way to remove all these copy-paste and factorize some code?

Reply:
I removed the calls from the tools being edited in this PR but quite a few remain elsewhere...

Reply:
Turns out one of the pipelines uses an old GNU compiler that does not support the filesystem include. I'll need to keep the old version in place for now.